tests/sharedtria/tria_02.cc

```diff
@@ -0,0 +1,115 @@
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+
+// create a shared tria mesh with artificial cells and refine it
+
+#include "../tests.h"
+#include <deal.II/base/logstream.h>
+#include <deal.II/base/tensor.h>
+#include <deal.II/grid/tria.h>
+#include <deal.II/distributed/shared_tria.h>
+#include <deal.II/grid/tria_accessor.h>
+#include <deal.II/grid/tria_iterator.h>
+#include <deal.II/grid/grid_generator.h>
+#include <deal.II/grid/grid_out.h>
+#include <deal.II/dofs/dof_handler.h>
+#include <deal.II/numerics/data_out.h>
+
+#include <fstream>
+
+template <int dim, int spacedim>
+void write_mesh (const parallel::shared::Triangulation<dim,spacedim> &tria,
+                 const char                                *filename_)
+{
+  DataOut<dim> data_out;
+  data_out.attach_triangulation (tria);
+  Vector<float> subdomain (tria.n_active_cells());
+  typename  parallel::shared::Triangulation<dim>::active_cell_iterator it=tria.begin_active();
+  for (unsigned int i=0; it!=tria.end(); ++it,++i)
+    subdomain(i) = it->subdomain_id();
+
+  data_out.add_data_vector (subdomain, "subdomain");
+
+  data_out.build_patches ();
+  const std::string filename = (filename_ +
+                                Utilities::int_to_string
+                                (tria.locally_owned_subdomain(), 4));
+  {
+    std::ofstream output ((filename + ".vtu").c_str());
+    data_out.write_vtu (output);
+  }
+}
+
+
+
+template<int dim>
+void test()
+{
+  parallel::shared::Triangulation<dim> tr(MPI_COMM_WORLD,
+                                          Triangulation<dim>::none,
+                                          /*artificial*/true);
+
+
+  GridGenerator::hyper_cube(tr);
+  tr.refine_global();
+  tr.begin_active()->set_refine_flag();
+  tr.execute_coarsening_and_refinement ();
+  tr.begin_active()->set_refine_flag();
+  tr.execute_coarsening_and_refinement ();
+
+  deallog
+      << " locally_owned_subdomain(): " << tr.locally_owned_subdomain() << "\n"
+      << " n_active_cells: " << tr.n_active_cells() << "\n"
+      << " n_levels: " << tr.n_levels() << "\n"
+      << " n_global_levels: " << tr.n_global_levels()  << "\n"
+      //<< " n_locally_owned_active_cells: " << tr.n_locally_owned_active_cells() << "\n"
+      //<< " n_global_active_cells: " << tr.n_global_active_cells() << "\n"
+      << std::endl;
+
+  /*deallog << "n_locally_owned_active_cells_per_processor: ";
+  std::vector<unsigned int> v = tr.n_locally_owned_active_cells_per_processor();
+  for (unsigned int i=0;i<v.size();++i)
+    deallog << v[i] << " ";
+    deallog << std::endl;*/
+
+  deallog << "subdomains: ";
+  typename  parallel::shared::Triangulation<dim>::active_cell_iterator it=tr.begin_active();
+  for (; it!=tr.end(); ++it)
+    {
+      deallog << it->subdomain_id() << " ";
```

Comment:
This outputs the awkward number `4294...` for `invalid_subdomain_id`. Can you write this as
```
  deallog << (int)it->sub....
```
to have it printed as `-1` so that we don't rely on the exact data type?